File: hangman/Kotlin/hangman.kt
import java.io.File
import java.io.InputStream
import java.util.concurrent.ThreadLocalRandom
import java.util.Scanner

fun main(args: Array<String>) {
    val w = mutableListOf<String>()
    val d = hashMapOf<Int, MutableList<String>>()
    val inputStream: InputStream = File("resources/words.in").inputStream()

    try {
        inputStream.bufferedReader().useLines { lines -> lines.forEach { w.add(it.trim())} }

        for(e in w) {
            if(!d.containsKey(e.length)) {
                d.put(e.length, mutableListOf())
            }
            d.get(e.length)!!.add(e)
        }

        print("Number of letters in the word? ")

        val l = readLine()!!.toInt()
        
        if(d.containsKey(l)) {
            val word = d.get(l)!!.get(ThreadLocalRandom.current().nextInt(0, d.get(l)!!.size))

            val v = BooleanArray(l)
            var e = 0
            var done = true

            while(e < 10) {
                done = true
                for (bool_val in v) {
                    if (!bool_val) {
                        done = false
                    }
                }
                if (done) {
                    break
                }

                print("Guess a letter: ")

                val chr = Scanner(System.`in`).nextLine().get(0)

                var hit = false

                for(i in 0..(word.length-1)) {
                    if(word.get(i) == chr && !v[i]) {
                        v[i] = true
                        hit = true
                    }
                }

                if (hit) {
                    println("Hit!")
                } else {
                    println("Missed, mistake #${e+1} out of 10")
                    
                    ++e

                    // drawing hangman
                    println()
                    if(e > 2) println("    xxxxxxxxxxxxx")
                    else println()

                    if(e > 3) println("    x           x")
                    else if (e > 1) println("                x")
                    else println()

                    if(e > 3) println("    x           x")
                    else if (e > 1) println("                x")
                    else println()

                    if(e>4) println("   xxx          x")
                    else if (e > 1) println("                x")
                    else println()

                    if(e>4) println("  xxxxx         x")
                    else if (e > 1) println("                x")
                    else println()

                    if(e>4) println("   xxx          x")
                    else if (e > 1) println("                x")
                    else println()

                    if(e>5) println("    x           x")
                    else if (e > 1) println("                x")
                    else println()

                    if(e>7) println("  x x x         x")
                    else if(e>6) println("  x x           x")
                    else if(e>5) println("    x           x")
                    else if (e > 1) println("                x")
                    else println()

                    if(e>7) println(" x  x  x        x")
                    else if(e>6) println(" x  x           x")
                    else if(e>5) println("    x           x")
                    else if (e > 1) println("                x")
                    else println()

                    if(e>5) println("    x           x")
                    else if (e > 1) println("                x")
                    else println()

                    if(e>9) println("   x x          x")
                    else if(e>8) println("   x            x")
                    else if (e > 1) println("                x")
                    else println()

                    if(e>9) println(" x     x        x")
                    else if(e>8) println(" x              x")
                    else if (e > 1) println("                x")
                    else println()

                    if(e>1) println("                x")
                    else println()

                    if(e>1) println("                x")
                    else println()

                    if(e>0) println("xxxxxxxxxxxxxxxxxxxxxxxxxxxx")
                    else println()
                    println()

                }
                print("The word: ")

                for(i in 0..(word.length-1)) {
                    if (v[i]) {
                        print(" " + word.get(i)+ " ")
                    } else {
                        print(" _ ")
                    }
                }

                println("\n")
            }

            // printing final result
            if (done) {
                println("You won!")
            } else {
                println("You lost.")
                println("The word was: ${word}\n")
            }

        } else {
            println("Sorry, no words like that.")
        }
    } catch (ex: Exception) {
        ex.printStackTrace()
    }
}

File: hangman/Lua/hangman.lua
local allWords = {}
local letters = {}

function load_words()
    local file = assert(io.open("resources/words.in", "r"))

    for word in file:lines() do
        table.insert(get_words_array(string.len(word)), word)
    end

    file:close()
end

function get_words_array(wordLength)
    local wordsArray = allWords[wordLength]
    wordsArray = type(wordsArray) == "table" and wordsArray or {}
    allWords[wordLength] = wordsArray
    return allWords[wordLength]
end 

function print_guess()
    local output = ""
    for _, v in ipairs(letters) do
        if v[2] == true then
            output = output .. " " .. v[1]
        else
            output = output .. " _"
        end
    end
    print("The word:" .. output)
end

function guess(guessed)
    local hit = false
    for _, v in ipairs(letters) do
        if v[1] == guessed then
            v[2] = true
            hit = true
        end
    end
    return hit
end

function has_guessed_all_letters()
    for _, v in ipairs(letters) do
        
        if v[2] == false then 
             return false
        end
    end

    return true
end

function print_hangman(mistakes)
    print()

    if mistakes > 2 then print("    xxxxxxxxxxxxx")
    else print() end

    if mistakes > 3 then print("    x           x")
    elseif mistakes > 1 then print ("                x")
    else print() end

    if mistakes > 3 then print("    x           x")
    elseif mistakes > 1 then print ("                x")
    else print() end

    if mistakes > 4 then print("   xxx          x")
    elseif mistakes > 1 then print ("                x")
    else print() end

    if mistakes > 4 then print("  xxxxx         x")
    elseif mistakes > 1 then print ("                x")
    else print() end

    if mistakes > 4 then print("   xxx          x")
    elseif mistakes > 1 then print ("                x")
    else print() end

    if mistakes > 5 then print("    x           x")
    elseif mistakes > 1 then print ("                x")
    else print() end

    if mistakes > 7 then print("  x x x         x")
    elseif mistakes > 6 then print ("  x x           x")
    elseif mistakes > 5 then print ("    x           x")
    elseif mistakes > 1 then print ("                x")
    else print() end

    if mistakes > 7 then print(" x  x  x        x")
    elseif mistakes > 6 then print (" x  x           x")
    elseif mistakes > 5 then print ("    x           x")
    elseif mistakes > 1 then print ("                x")
    else print() end

    if mistakes > 5 then print("    x           x")
    elseif mistakes > 1 then print ("                x")
    else print() end

    if mistakes > 9 then print("   x x          x")
    elseif mistakes > 8 then print ("   x            x")
    elseif mistakes > 1 then print ("                x")
    else print() end

    if mistakes > 9 then print(" x     x        x")
    elseif mistakes > 8 then print (" x              x")
    elseif mistakes > 1 then print ("                x")
    else print() end

    if mistakes > 1 then print ("                x")
    else print() end

    if mistakes > 1 then print ("                x")
    else print() end

    print ("xxxxxxxxxxxxxxxxxxxxxxxxxxxx")
    print()
end

function play()
    load_words()

    print("Number of letters in the word?")
    local key = tonumber(io.read())

    local words = get_words_array(key)

    if words[1] == nil then
        print("Sorry, no words like that.")
        return
    end

    -- load a random word
    math.randomseed(os.time())
    local word = words[math.random(#words)]

    -- put word into table
    for c in word:gmatch(".") do
        table.insert(letters, {c , false})
    end

    -- play the game
    local mistakes = 0
    local done = false
    while not(done) do

        print("Guess a letter:")
        local guessedLetter = io.read():lower()
        
        if guess(guessedLetter) then
            print("Hit")
        else
            mistakes = mistakes + 1
            print("Missed, mistake " .. mistakes .. " out of 10")

            print_hangman(mistakes)
        end

        print_guess()

        if has_guessed_all_letters() then
            print("You won!")
            return
        end

        if mistakes >= 10 then
            print("You lost.")
            print("The word was: " .. word)
            return
        end

    end

end

play()
